lib/saluki-components/src/sources/dogstatsd/mod.rs

```diff
@@ -392,108 +429,181 @@ async fn process_listener(source_context: SourceContext, listener_context: Liste
 
-        match deserializer.decode(&mut event_buffer).await {
-            // No events decoded. Connection is done.
-            Ok((0, peer_addr)) => {
-                trace!(%listen_addr, %peer_addr, "Stream received EOF. Shutting down handler.");
-                break;
-            }
-            // Got events. Forward them.
-            Ok((n, peer_addr)) => {
-                // We do one optional enrichment step here, which is to add the client's socket credentials as a tag
-                // on each metric, if they came over UDS. This would then be utilized downstream in the pipeline by
-                // origin enrichment, if present.
-                if origin_detection {
-                    if let ConnectionAddress::ProcessLike(Some(creds)) = &peer_addr {
-                        for event in &mut event_buffer {
-                            if let Some(metric) = event.try_as_metric_mut() {
-                                metric
-                                    .metadata_mut()
-                                    .origin_entity_mut()
-                                    .set_process_id(creds.pid as u32);
-                            }
-                        }
-                    }
-                }
-
-                // Extract eventd events only if at least one is present in the event buffer.
-                let maybe_eventd_event_buffer = match event_buffer.has_data_type(DataType::EventD) {
-                    true => {
-                        let mut eventd_event_buffer = source_context.event_buffer_pool().acquire().await;
-                        eventd_event_buffer.extend(event_buffer.extract(Event::is_eventd));
-                        Some(eventd_event_buffer)
-                    }
-                    false => None,
-                };
-
-                // Extract service check events only if at least one is present in the event buffer.
-                let maybe_service_checks_event_buffer = match event_buffer.has_data_type(DataType::ServiceCheck) {
-                    true => {
-                        let mut service_check_event_buffer = source_context.event_buffer_pool().acquire().await;
-                        service_check_event_buffer.extend(event_buffer.extract(Event::is_service_check));
-                        Some(service_check_event_buffer)
-                    }
-                    false => None,
-                };
+        let mut buffer = io_buffer_pool.acquire().await;
+        debug!(capacity = buffer.remaining_mut(), "Acquired buffer for decoding.");
 
-                trace!(%listen_addr, %peer_addr, events_len = n, "Forwarding events.");
+        // If our buffer is full, we can't do any reads.
+        if !buffer.has_remaining_mut() {
+            // return Err(DeserializerError::BufferFull {
+            //     remaining: buffer.remaining(),
+            // });
+            // try to get a new buffer on the next iteration?
+            continue;
+        }
```

Comment:
This should also throw an error log, because a fresh I/O buffer should always be cleared after acquiring it.

Reply:
Yeah this is a bit of an odd one. Part of me wants to panic because if it happens we have a logic bug somewhere, the other part thinks it's a non-issue and we can just clear the buffer 😅 .

For now I'll just add the error log.